### DISCUSSION TITLE
1.18.0

### DISCUSSION THREAD

**Discussion by attipaci** [nom-tam-fits/nom-tam-fits#465] [Announcements]
## Summary

This is a big feature release, fixing tile compression and many issues with binary tables, packing several cool new features, and numerous improvements. Many thanks to Dustin Jenkins and Kevin Eastridge for their contributions to making this the best nom-tam-fits yet. 

## Bugs Fixed

 - #349 -- Tiled GZIP (de)compression of quantized images now works!
 - #377 -- Use of `ZDITHER0` keyword to store or restore the random generator seed when dithering is used in compression or decompression.
 - #436 -- Fixed incorrect handling of `CONTINUE` keywords with no quoted string value (a regression since 1.16).
 - #438 -- Conversion of variable length `float[]` or `double[]` arrays to complex was broken since inception. Now it works for the first time as expected.
 - #455 -- Fixed incorrect variable-length complex array descriptors in binary tables.
 - #445 -- Consistent `String` return for both fixed and variable-length string columns in binary tables.
 - #456 -- Fixed compressed table tiling with tile sizes that do not divide the table rows.

## New features

 - #387 -- Can generate sparsely sampled tiles from large images via new striding feature.
 - #356 and #400 -- Streaming of selected cutouts from large images, including from compressed images.
 - #352 -- Alternative random access implementations via `RandomAccessIO`, providing file-like access e.g. to cloud resources (such as Amazon S3)
 - #335 -- New `NullData` and `NullDataHDU` classes for more intuitive creation of HDUs with headers only, such as dummy primary HDUs before extensions.
 - #183 -- Added support for `ComplexValue`-based column in binary tables.
 - #450 -- Added support for bit-based columns (FITS type `X`) in binary tables.
 - #451 -- Added support for `null` (undefined) logical values for FITS logical columns in binary tables.
 - #448 -- More user-friendly data access in `BinaryTable`, using Java boxing for scalar primitive elements, and automatic type conversion (both narrowing and widening) for singleton values when possible. Many new `BinaryTable` methods to make it easier to build and use binary tables.
 - #437 -- Binary table access and editing directly in file, when getting or setting table elements in deferred-read mode.
 - #453 -- Allow defragmenting binary table heaps to expunge stale data, and to optimize heap area for best sequential read performance.
 - #452 -- Simplified access to select tiles / regions of compressed data in both tile compressed tables and images.

## Improvements

 - #414 -- Significantly improved IO performance, especially for images. Image HDUs now read/write __2.5x -- 4x__ faster than in earlier versions, and read a whopping __25x -- 50x__ faster from streams than in __1.15.2__.
 - #407 -- Improved dithering speed, by calculating random sequence only once when needed, and doing it more efficiently.
 - #441 -- Significantly faster table data access and row additions.
 - #447 -- Check data consistency when adding or setting data in binary tables.
 - #454 -- Do not use ASCII NUL (`0x00`) for padding between string array elements in binary tables, since it violates the FITS standard (according to which an ASCII NUL terminates the complete string array entry).
 - #131 -- Tables now support up to ~2-billion (`Integer.MAX_VALUE`) rows regardless of entry size.
 - #207 -- Unified deferred read for all `Data` implementations. This brings deferred reading to `AsciiTable`, `RandomGroups`, and `UndefinedData` classes for the first time. 
 - #395 -- Consistent behavior of all `Data` subclasses. _Because they has subtle differences in the past, the change may break some applications, although it's more likely that the gained uniformity will eliminate sneaky bugs that may have lurked in any affected applications._
 - #426 -- No longer exposing critical internal arrays to users, giving them copies instead, which can be modified without breaking functionality of the library.
 - #435 -- Header access methods in `BasicHDU` no longer remove significant leading spaces from strings to better conform to FITS specification.
 - #394 -- Overhauled `FitsDate` to make it simpler and better, without changing functionality.
 - #424 -- Consistent source code formatting etc.

## API changes

 - #446 -- Disable automatic (and inconsistent) detection of complex binary table data, when adding real-valued data columns to a binary table. Users should call `BinaryTable.setComplexColumn()` explicitly if pairwise real values should be converted to complex. 
 - #434 -- Stop the automatic creation of random group HDUs from compatible table data. Instead create binary or ASCII tables as appropriate given the `FitsFactory` settings. The FITS standard discourages the use of random group HDUs going forward, and support should be typically limited to dealing with some older existing data files in that format. If there is ever a need to create new random group HDUs, you can do it explicitly via `RandomGroupsHDU.createFrom(Object[][])`. 
 - #209 -- Deprecated `FitsUtil.reposition()` and changed argument type to `FitsIO`, which is more appropriate for its use.
 - #420 -- Revised `PaddingException` API and seamless internal handling.
 - #336 -- Deprecated a number of unsafe methods that should be used internally only (if at all). These should never have been exposed to users in the first place.
 - #425 -- Deprecated the use of hexadecimal values in headers. FITS does not support these, and you should not use them.
 - #430 -- Deprecated `BigInteger` and `BigDecimal` values in headers. FITS does not support these, and you should not use them.
 - #429 -- Removed unnecessary IO synchronization from `FitsFile`, `FitsInputStream`, and `FitsOutputStream` classes.

## Documentation

 - Fully revised [README](https://github.com/nom-tam-fits/nom-tam-fits/blob/master/README.md) to provide an up-to-date introduction to the library's features, including more and better usage examples.   
 - #427 -- Online and packaged API  documentation is restricted to public classes and methods only, as appropriate for users who want to read/write FITS files. Classes and methods intended for internal use are clearly indicated when these are exposed in the public API. Contributors can still generate the full documentation locally if desired using maven.
 - #160 -- Completed API docs for classes and methods intended for users.

<hr /><em>This discussion was created from the release <a href='https://github.com/nom-tam-fits/nom-tam-fits/releases/tag/1.18.0'>1.18.0</a>.</em>

**Comment by at88mph** [comment]
🎉 
Thank you very much @attipaci!  Will this go to Maven Central today?

**Reply by attipaci** [reply]
It's been submitted, but it might take up to 48 hours before Maven Central picks it up.
